### ISSUE TITLE
Handling errors in actions

### ISSUE THREAD

**Issue by SteffenDE** [Cantido/liberator#33]
Hi there,

I really like the idea of liberator, but while trying it in a project of mine I came across the following issue and I do not know how to properly move forward:

When creating a new resource, I implemented the `post!` action, but I cannot find a nice way to handle errors happening during creation, e.g. a unique check fails that is only first recognized when Ecto tries to insert into the database.

```elixir
@impl true
def post!(conn) do
    with {:ok, resource} <- MyApp.Resource.create_resource(conn.params) do
        # returning something from an action seems to be a no-op
        %{resource: resource}
    else
        %{error: %Ecto.Changeset{}} -> # how do I handle this?
    end
end
```

When working with plain Phoenix one usually has a FallbackController that handles Ecto changeset errors. Maybe one could allow returning values from actions, like returning the conn struct to end processing immediately. Then one could do something like this:

```elixir
@impl true
def post!(conn) do
    with {:ok, resource} <- MyApp.Resource.create_resource(conn.params) do
        %{resource: resource} # merge this into the assigns to reference it in handle_created
    else
        other -> MyApp.FallbackController.call(conn, other)
    end
end
```

I realize that validating the parameters could be normally done in the `well_formed?` step, but when the error only happens on insert this cannot easily be done.
What do you think?

In any case, thanks for this great library! It's a mystery to me why it has so little attention.

**Comment by Cantido** [comment]
Interesting proposal! While developing this I was kinda assuming that Phoenix would handle the error with a 500, and that would be fine, but you're totally right, having fallback controller functionality would be really preferable. Right off the top of my head, I'm thinking about adding another `handler_*` function to handle errors like that. Let me noodle on this for a bit.

I also have a lot of changes unreleased that I need to get around to cutting a release for, I think better error handling would a great addition.

**Comment by SteffenDE** [comment]
Yes, Phoenix does respond with a 500, but that's normally not how I want to handle errors. So yeah I guess that a `handle_error` function would be great. Despite that, I think it would be nice to be able to return values from actions that get merged into the assigns as with decisions:

```elixir
@impl true
def post!(conn) do
    new_resource = MyApp.Resource.create(conn.params)
    %{created_resource: new_resource}
end

@impl true
def handle_created(conn) do
    conn.assigns.created_resource
end
```

This currently does not work and therefore one would need to fetch the created resource again to return it, I guess.

I'll be following the progress here and in the meantime continue to write "raw" Phoenix Controllers.

**Comment by Cantido** [comment]
That makes perfect sense. The only reason those actions ignore the return value is because I think that's what the Clojure library did, so my reasoning wasn't very strong there :)

**Comment by Cantido** [comment]
I have good news! There's currently an unreleased change that lets all the handlers return values just like the rest of the decision functions. I had forgotten that.

Also, I'm working on the error handling bit now. I'll make sure it behaves just like a fallback controller. Let me know if there's anything else you want included. Once I add this, I'll release an RC version since I currently have a lot of unreleased changes I want to let settle. Check out the changelog, I'd love your feedback on what I've got in the pipeline! I haven't had any feedback on this library yet.

**Comment by Cantido** [comment]
I've just added a `handle_error(conn, error, failed_step)` callback, and I published it to hex with the version `1.4.0-rc.1`.  You can add an RC version just like this:

    {:liberator, "~> 1.4.0-rc.1"}

The added docs for this function are right here: https://hexdocs.pm/liberator/1.4.0-rc.1/Liberator.Resource.html#module-handling-errors.

Please give that a try and let me know what you think!

**Comment by SteffenDE** [comment]
Thank you for the quick update. I tried to implement a fallback-controller like behaviour in an example repo using Phoenix + Liberator. I didn't use the new `handle_error` callback as I cannot easily raise a changeset, but rather used the fact that one can return values from handlers now. Returning a conn does not seem to work, as I get a `(Protocol.UndefinedError) protocol Jason.Encoder not implemented for %Plug.Conn{...}` error.

I uploaded the code over here: https://github.com/SteffenDE/liberator_phoenix_api
I tried to keep it very simple. The error is thrown when trying to post against `/api/myresource`.

I'd be really cool if one could use the normal Phoenix FallbackController, but it uses the controller's `render` function which itself calls `send_resp`.
Sidenote: when changing `resp` to `send_resp` in my `FallbackResource` [module](https://github.com/SteffenDE/liberator_phoenix_api/blob/b06c2b548ac90ff0bd0c6ddfa7657da54026345f/lib/liberator_test_web/controllers/fallback_resource.ex#L8) the wanted response is sent and afterwards the same `Protocol.UndefinedError` as above gets thrown.

As an improvement I'd really like to see an example of implementing an api resource with Phoenix + Ecto in the docs. Once I figured out the best way to do this I'd be happy to contribute a section about this to the documentation.

**Comment by Cantido** [comment]
I'll see if I can get a fallback controller wired up. I think I can get around the `send_resp` issue.

It's been on my to-do list to add a more detailed example of an API resource. I'd love to evolve this library as we refine what a proper API implementation should look like. I wrote a really basic example resource when I put together the Getting Started document, and that informed quite a few changes I made for the current version.

**Comment by Cantido** [comment]
Also I realized I was wrong in an earlier comment. All handler functions except `handle_error` still need to return the response body. It's the actions like `post!` that can now return a `conn` or a map in this updated version.

Your example looks great. I think though at [line 32](https://github.com/SteffenDE/liberator_phoenix_api/blob/main/lib/liberator_test_web/controllers/my_resource.ex#L32) there needs to be a call to Ecto. Something like this would throw an exception that the handler could catch:

```elixir
  def post!(conn) do
    LiberatorTest.Resource.changeset(%LiberatorTest.Resource{}, %{"email" => "abc"})
    |> LiberatorTest.Repo.insert!()
  end
```

**Comment by SteffenDE** [comment]
I updated the code and it works now, see https://github.com/SteffenDE/liberator_phoenix_api/commit/cc43046e82c5ffcc022245e32802427b1153eb24. Still I'm not sure if I always want to raise an error. Often one already has context functions that return `{:ok, data}` and `{:error, error}` so the transition is not as smooth as it could be when moving from a Phoenix controller to Liberator.

**Comment by Cantido** [comment]
Yeah I gotcha. I've been considering responding to error tuples. I think I'll go ahead with that.

**Comment by Cantido** [comment]
Next version is ready. This one includes:
- you now need to call `send_resp` inside the error handler, so you can re-use a fallback controller
- All decisions, actions, and handlers can return error tuples. Returning `{:error, result}` will execute `handle_error`. Returning `{:ok, result}` will act as if you returned `result`, which can be a map, or any truthy value.

It's `rc.3` because I merged a dependabot update after I tagged `rc.2` and I wanted to make sure I was tagging the same version I uploaded to hex.

```elixir
{:liberator, "~> 1.4.0-rc.3"}
```

**Comment by SteffenDE** [comment]
This is awesome! I've updated my demo code and added a default Phoenix FallbackController and it works flawlessly. This also allowed me to simplify the `exists?` decision, but I guess that this is not the optimal way as raising there skips some steps in the decision tree, right?
In any case I really like these changes. This makes the interoperability in a Phoenix project really easy.

**Comment by Cantido** [comment]
I'm so happy to hear that! Yeah exceptions and error tuples will always go straight to `handle_error`. Returning `false` in the `exists?` decision goes directly to `handle_not_found`, so you're not skipping any steps, you're just missing out on the handler made for that result.

**Comment by Cantido** [comment]
I just released version 1.4.0 with these changes. Thank you so much for helping me improve this library! If you run into any other problems, please create an issue here; I really appreciate the feedback.

**Comment by Cantido** [comment]
One more thing; I finally realized that there is a decision callback called `conflict?/1` that is a good place to check for data conflicts like you mentioned in the issue, it's called before PUT and POST requests. I know you said that you don't know if there is a conflict until you try to update the DB, but I thought I'd mention it. If you wanted to implement the additional check, you'd win the ability to use the `handle_conflict` handler and status code.
